Community review

JayPanoz opened the pull request:
This branch should make it easier for the community to review.

JayPanoz committed:
Init Publication Encapsulation

JayPanoz committed:
Add package.json + gitignore for node_modules

JayPanoz committed:
Temporarily disable link parameters (RFC6570)

JayPanoz committed:
Add parameters and encoding to MediaType

JayPanoz committed:
Improve structuredSyntaxSuffix() method

JayPanoz committed:
Make corrections in Link and Publication

JayPanoz committed:
Add missing title in Link constructor

JayPanoz committed:
Correct Links class

JayPanoz committed:
Correct loop in MediaType constructor

JayPanoz committed:
Improve Links constructor

JayPanoz committed:
Correct contains in MediaType

JayPanoz committed:
Handle number in Links constructor

JayPanoz committed:
Add format class

JayPanoz committed:
Add format + formatVersion in Publication

JayPanoz committed:
Add type in Metadata

JayPanoz committed:
Switch back to original getManifest() method

This static method could be highly opinionated as there are various
strategies one can use so let’s start from the original one.

JayPanoz committed:
Add missing R2D2BC methods in PublicationManifest

JayPanoz committed:
Switch back to URL for manifestUrl to minimise breakage

JayPanoz committed:
Add OPDS in MediaTypes and Formats

JayPanoz committed:
Correct type of manifestURL in Divina and EPUB Manifests

JayPanoz committed:
Switch back to mandatory manifestURL in Publication Manifest

mickael-menu commented on r2-shared/Format/Format.ts:
```diff
@@ -0,0 +1,147 @@
+  name: string;
+  mediaType: MediaType;
+  fileExtension: string;
+}
```
Bear with me, because I don't know much about TypeScript, but why use an interface in this case? `Format` will have only one implementation.

I'm also a bit confused how `Format` is created in its constructor, why not `constructor(name: string, mediaType: MediaType, fileExtension: string)`?

JayPanoz replied:
It’s more like I wanted to keep named parameters in static methods for more clarity as this can turn super ugly when done directly in `constructor`, cf. this [SO question](https://stackoverflow.com/questions/45507428/typescript-create-class-via-constructor-passing-in-named-parameters).

mickael-menu replied:
@danielweck During the call there was a consensus in favor of using interfaces to create anonymous object arguments to replace the concept of "named parameters". But we're not sure about the naming of such interfaces. Is there a standard convention for this? Or a more idiomatic way to tackle the issue of named parameters?

mickael-menu replied:
How about using a type alias on the object structure instead of an interface? This doesn't solve the naming issue, but at least it's clear that it's not a type to be implemented.

```typescript
type FormatDefinition = { name: string }

class Format {
  public name: string;

  constructor(params: FormatDefinition) {
    this.name = params.name;
  };

}
```

A few other ideas for the name: `FormatDefinition`, `FormatConstructor`, `FormatSchema`, `FormatData` and `FormatLike`, as seen in this PR.

Technically we could also ditch the name entirely. But I don't know enough TypeScript to see if that could be a problem:

```typescript
class Format {
  public name: string;

  constructor(params: { name: string }) {
    this.name = params.name;
  };

}
```

danielweck replied:
I have come across a number of Javascript / TypeScript APIs where objects are used as function parameters in order to pass multiple atomic pieces of data (quick example from the top of my head: React component props and state), but there are many instances of the opposite design pattern (extreme example: curry-ing which “spreads” individual arguments over multiple function calls).

I think that generalising the use of objects to carry multiple function arguments is an unusual approach, but as far as I know it is not an anti-pattern ... though I am sure quite a few JS/TS programmers will find this “weird”, so I am not convinced this will improve the developer experience (DX) across the board.

Note that there is a runtime cost incurred when creating objects instead of just spreading the function arguments the “normal” way. Garbage collection in JS engines is pretty good, and the Readium SDK is not a “real time” / “high performance” system ... but still, this is something to keep in mind in addition to DX considerations.

Another potential problem is that variable with primitive types passed as function arguments are sometimes not used as constants within the closure scope, and instead are modified directly in order to avoid creating unnecessary stack allocations (i.e. this is an implementation detail below the API surface). With objects, there is the potential problem of mutating properties in one function scope, and inadvertently changing the value in another. Just a thought.

mickael-menu replied:
Thanks for chiming in @danielweck. Yeah that would probably be overkill. If I understood correctly, @JayPanoz planned to use these mostly for large constructors with multiple parameters of the same type. One could argue that `Format` doesn't fit the bill, but it would probably be safer for `Link`, `Metadata`, etc. Because adding a new parameter constructors in an update means that the constructor could break silently if the type of the new parameter is the same.

danielweck replied:
Yes, it is quite common in JavaScript/TypeScript APIs to use a "config" object argument which properties are assigned defaults when not defined on the caller's side. I can see how this would work for constructors or other initializers. That being said, it is also quite common to have such "config" function argument alongside other arguments, so the design pattern is quite open.

Personally I would favour consistency, so if "object with named properties is used to pass function arguments" gets applied to constructors, then I would prefer the design pattern to be used for all other public API functions (which then becomes quite cumbersome, as noted in a previous comment).

mickael-menu commented on r2-shared/Format/MediaType.ts:
```diff
@@ -0,0 +1,269 @@
+  public parameters: ParametersMap;
+
+  constructor(mediaType: string) {
+    const components = mediaType.replace(/\s/g, "").split(";");
+    const types = components[0].split("/");
+    if (types.length === 2) {
+      this.type = types[0].toLowerCase();
+      this.subtype = types[1].toLowerCase();
+    }
+
+    let parameters: ParametersMap = {};
+    for (let i = 1; i < components.length; i++) {
+      const component = components[i].split("=");
+      if (component.length === 2) {
+        const key = component[0];
+        const value = component[1];
+        parameters[key] = value;
+      }
+    }
+    this.parameters = parameters;
+
+    let parametersString: string = "";
+    for (const p in parameters) {
+      const value = parameters[p];
+      parametersString += `;${p}=${value}`;
+    }
+    this.string = `${this.type}/${this.subtype}${parametersString}`;
+
+    this.encoding = parameters["encoding"];
```
This part is quite complicated, so I think it would be important to add some unit tests once we have the test suite set up, [based on Swift for example](https://github.com/readium/r2-shared-swift/blob/develop/r2-shared-swiftTests/Format/MediaTypeTests.swift).

Also for the comparison functions below.

mickael-menu commented on r2-shared/Format/MediaType.ts:
```diff
@@ -0,0 +1,269 @@
+    return parts.length > 1 ? `+${parts[parts.length - 1]}` : null;
+  }
+
+  public contains(other: MediaType | string): boolean {
```
It would be useful to have doc comments above each public function. You can use the ones from Swift or Kotlin, or directly from the proposal which they are derived from.

(Applies to the whole PR)

mickael-menu commented on r2-shared/Format/MediaType.ts:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  public matches(other: MediaType | string): boolean {
+    return this.contains(other);
```
`matches()` is not just an alias to `contains()`, but needs to work in both directions. So it should be something like this:

```typescript
return this.contains(other) || other.contains(this)
```

But I think the focus should be on implementing unit tests, they should catch that kind of errors.

JayPanoz replied:
Note: Resolve Pending as I wanted to doublecheck.

mickael-menu commented on r2-shared/Format/Format.ts:
```diff
@@ -0,0 +1,147 @@
+    this.fileExtension = format.fileExtension;
+  };
+
+  public static audiobookManifest(): Format {
```
It's not yet updated in the `Format` proposal, but [following this discussion](https://github.com/readium/r2-streamer-kotlin/pull/105#discussion_r433109026) we thought it would be better to prefix `audiobook` and `webpub` formats/media types with `readium`, to disambiguate with W3C flavors. This doesn't impact DiViNa.

So: `readiumAudiobook`, `readiumAudiobookManifest`, `readiumWebPub`, `readiumWebPubManifest`

mickael-menu commented on r2-shared/Format/Format.ts:
```diff
@@ -0,0 +1,147 @@
+    });
+  }
+
+  public static webpub(): Format {
```
By the way here you're using the `webpubManifest` media type, but the name of the format is `webpub`. Shouldn't it be `readiumWebPubManifest` instead? And is the shared TypeScript layer supposed to handle WebPubs packages? If yes, then we also need to add the formats and media types without the "manifest" suffix.

mickael-menu commented on r2-shared/Format/Format.ts:
```diff
@@ -0,0 +1,147 @@
+      mediaType: MediaType.webpubManifest(),
+      fileExtension: "json"
+    });
+  }
```
Some formats are missing, but maybe they are not necessary on the Web platform? I think it's best to implement just what's needed.

JayPanoz replied:
So that is a good question. 

I was not sure how to handle these missing ones (same with media types). Didn’t want to imply packages/ZIP are handled as it was discussed long ago that if someone wanted to support these, they would have to make their own module.

mickael-menu replied:
If there's a potential to support packages, I think we should at least add them to the media types and formats but leave the implementation to third-parties. This way we control that the proper media type, extension and name is used.

mickael-menu commented on r2-shared/Format/MediaType.ts:
```diff
@@ -0,0 +1,269 @@
+    return new this("audio/aiff");
+  }
+
+  public static audiobookManifest(): MediaType {
```
The media types for packaged WebPub variants are missing, [they are listed in the proposal](https://github.com/readium/architecture/blob/master/proposals/001-format-api.md#constants).

mickael-menu commented on r2-shared/Publication/Locator.ts:
```diff
@@ -0,0 +1,34 @@
+  charOffset?: number;
+}
+
+interface DomRangeInfo {
```
So I'm not sure what's the convention for TypeScript, to check with the team. But we need to be consistent with the capitalization of acronyms. If it's based on JavaScript, I guess it should be `DOMRangeInfo`.

mickael-menu replied:
Also the name of these structures is a bit outdated I think, it should be `DOMRange` : https://github.com/readium/architecture/blob/master/models/locators/extensions/html.md#the-start-and-end-object

for `StartEndInfo` there's no clear naming, but we used `DOMRange.Point` on mobile, since it fits the description: "A serializable representation of the "start" **boundary point** of the DOM Range"

mickael-menu commented on r2-shared/Publication/Contributor.ts:
```diff
@@ -0,0 +1,8 @@
+import { LocalizedString } from "./LocalizedString";
+
+export interface IContributor {
+  name: string | LocalizedString;
```
We talked a bit about this already, but on mobile we decided to always require a localized string, like that:

```swift
var localizedName: LocalizedString
var name: String { localizedName.string }
```

However, I think it would be fine like that too, depending on the idiom of the platform. Something to be discussed with the team?

JayPanoz replied:
Yeah `string` is there as a workaround to keep working on other things as LocalizedString has not been tackled yet.

mickael-menu replied:
Okay, thanks for the heads-up 👍

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+  series?: string | Collection;
+}
+
+type Contributor = string | IContributor | Array<string | IContributor>;
```
(This applies to actually almost all the types)

It looks a bit denormalized, in particular the array of contributors doesn't really fit a single type `Contributor`.

Personally I would be in favor of always using a full `Contributor` object, but that might not be idiomatic for TypeScript. I'll let more knowledgeable people chime in on this.

For example, when requiring `metadata.author`, we have to check if it's `null`, a `string`, a list of `string`, a `Contributor` or a list of `Contributor`. I guess this can get pretty unwieldy in the app. Consider instead having only a list of `Contributor` for `metadata.authors`. We know it can be empty, or we can rely on the fact that it contains only full-fledged `Contributor` objects.

JayPanoz replied:
Yeah I’d like to simplify that but need feedback from other parties.

mickael-menu replied:
Okay, let's keep this for Monday's meeting then.

JayPanoz replied:
I guess that from the feedback on Manifest, we can simplify that. I went ahead and simplified Contributor + Subject but we can make sure this won’t be an issue during Monday’s meeting.

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+  position?: number;
+}
+
+interface Collections {
```
This structure is not right (including `Metadata.belongsTo`), we should have a list of collections and a list of series, not a list of pairs (collection, series).

To simplify it can be done like that in `Metadata`:

```typescript
belongsToCollection: Array<Collection>
belongsToSeries: Array<Collection>
```

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+import { ReadingProgression } from "./ReadingProgression";
+import Presentation from "./presentation/Presentation";
+
+interface Collection extends IContributor {
```
Should this be prefixed with `I` if it's an interface?

JayPanoz replied:
Yeah I wish we can come up with best practices on when to use interfaces, type aliases, etc. and their syntax. 

Lots of people have been used to using interfaces by default because [you could not extend nor implement from type aliases prior to TS 2.7](https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases) so it can quickly become a mess if we don’t set guidelines.

mickael-menu commented on r2-shared/Publication/Contributor.ts:
```diff
@@ -0,0 +1,8 @@
+  name: string | LocalizedString;
+  sortAs?: string;
+  identifier?: string;
+  role?: string;
```
`links: Array<Link>` is missing

"Used to retrieve similar publications for the given contributor."

https://readium.org/webpub-manifest/schema/contributor-object.schema.json

mickael-menu commented on r2-shared/Publication/Encrypted.ts:
```diff
@@ -0,0 +1,7 @@
+export interface Encrypted {
```
We renamed this structure `Encryption` instead of `Encrypted` on mobile, it makes more sense IMHO.

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+  templated?: boolean;
+  type?: string;
+  title?: string;
+  rel?: string;
```
`rels` is actually a list of `string`

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+interface LinkProperties extends PresentationProperties {
+  contains?: Array<Properties>;
+  encrypted?: Encrypted;
+  mediaOverlay?: string;
```
I think that `mediaOverlay` doesn't exist here, is it legacy?

The extensibility is not supported, `link.properties` should actually just be a map `string -> any`, because all its members are actually part of RWPM modules: https://readium.org/webpub-manifest/schema/link.schema.json

I guess extension helpers could be implemented with something like that:

```typescript
LinkProperties.prototype.getEncryption = function() {
    // converts on-the-fly the JSON representation of `Encryption`
    return Encryption(this.properties["encrypted"])
}
```

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+  width?: number;
+  duration?: number;
+  bitrate?: number;
+  language?: string;
```
`languages` is actually a list of `string`

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+  duration?: number;
+  bitrate?: number;
+  language?: string;
+  alternate?: Array<ILink>;
```
We used `alternates` on mobile because it's a list.

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+  children?: Array<ILink>;
+}
+
+export class Link implements ILink {
```
Same comment as earlier, I don't understand the purpose of having an interface `ILink` here?

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+    } else {
+      return new URL(this.href, baseUrl).href;
+    }
+  }
```
I guess it could be more idiomatic to, [like Kotlin](https://github.com/readium/r2-shared-kotlin/blob/b3e4e59745a6cb2719437e418b3f8d45d85702b4/r2-shared/src/main/java/org/readium/r2/shared/publication/Link.kt#L99) call this `toURL(baseURL: string)` instead.

Also the behavior is not right: the `href` should be returned if it's already an absolute URL, and if it's relative, it should be appended to the given base URL.

This set of tests should help validate this function: https://github.com/readium/r2-shared-swift/blob/981594c6d935b8329a543b86b0a15a580ecc7508/r2-shared-swiftTests/Publication/LinkTests.swift#L218-L255

JayPanoz replied:
Went with `toAbsoluteHREF` cos’ in hindsight, you would expect an URL object in JS when using URL, cf. https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

Admittedly, the method could simply `return new URL(this.href, baseUrl).href` as base will be ignored if the link is already absolute – so it’s kinda built in.

mickael-menu replied:
On Swift I returned an `URL` object, but I don't know if it's the most convenient type for JS. Also I noticed that sometimes `href` contains invalid URL characters (especially for ZIP sources), so we need to make sure the path component is percent-encoded: https://github.com/readium/r2-shared-swift/blob/981594c6d935b8329a543b86b0a15a580ecc7508/r2-shared-swift/Publication/Link.swift#L124

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+  }
+
+  /*
+  public templateParameters(): Set<string> {
+    if (this.templated) {
+      
+    }
+  }
+
+  public expandTemplate(parameters: Array<string>): Link {
+    if (this.templated) {
+      
+    }
+    return this;
+  } */
+}
```
@qnga recently implemented a lightweight algorithm to handle most URI templates we're using.

Here's the code: 
* [in Link](https://github.com/readium/r2-shared-kotlin/blob/b3e4e59745a6cb2719437e418b3f8d45d85702b4/r2-shared/src/main/java/org/readium/r2/shared/publication/Link.kt#L77-L94)
* [the actual algorithm is abstracted in URITemplate](https://github.com/readium/r2-shared-kotlin/blob/b3e4e59745a6cb2719437e418b3f8d45d85702b4/r2-shared/src/main/java/org/readium/r2/shared/util/URITemplate.kt), because it's going to be useful outside of `Link` as well.
* [here are some unit tests to validate the parsing](https://github.com/readium/r2-shared-swift/blob/feature/publication-encapsulation/r2-shared-swiftTests/Toolkit/URITemplateTests.swift)

JayPanoz replied:
Thanks! 

Left that pending as I didn’t know if we would be using npm packages – and the available packages have massive shortcomings/limitations anyway.

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+    Object.setPrototypeOf(this, new.target.prototype);
+  }
+
+  public firstWithRel(rel: string): Link | null {
```
We should use the most idiomatic name for these functions on each platform, so for TypeScript it should be `findWithRel()` instead of `firstWithRel()`

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+    return this.find(predicate) || null;
+  }
+
+  public indexOfFirstWithHref(href: string): number {
```
Similarly, here we should use `findIndexWithHref()` to match Array's API

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  public firstWithMediaType(mediaType: string): Link | null {
+    const predicate = (el: Link) => el.type ? new MediaType(el.type).matches(mediaType) : false;
```
I think it could be useful to have `new MediaType(el.type)` in `Link` directly:

```typescript
mediaType?: MediaType = new MediaType(this.type)
```

mickael-menu commented on r2-shared/Publication/Link.ts:
```diff
@@ -0,0 +1,170 @@
+    return this.filter(predicate);
+  }
+
+  public allAreAudio(): boolean {
```
Similarly, to match Array's API, we can use `everyAreX` instead of `allAreX`

mickael-menu commented on r2-shared/Publication/LocalizedString.ts:
```diff
@@ -0,0 +1,3 @@
+export interface LocalizedString {
+  [key: string]: string
```
It would be useful to have an API that returns the string for the given locale (or the default user locale if none is given).

```typescript
toString(language?: string = null): string
```

mickael-menu commented on r2-shared/Publication/Locator.ts:
```diff
@@ -0,0 +1,34 @@
+}
+
+interface Locations {
+  fragments?: Array<string>;
```
It would be simpler to have `fragments` non-nullable. We can have an empty array when there're no fragments.

mickael-menu commented on r2-shared/Publication/Locator.ts:
```diff
@@ -0,0 +1,34 @@
+  totalProgression?: number;
+  cssSelector?: string;
+  partialCfi?: string;
+  domRange?: DomRangeInfo;
```
These ones should be added as helpers on the prototype. But we need `Locations` to support extensibility by adding a `otherLocations: Map<String, Any>`, for example.

JayPanoz replied:
Locator is in the same bag as Manifest actually, it first needs some feedback from other parties so I haven’t addressed it too much so far.

mickael-menu commented on r2-shared/Publication/Locator.ts:
```diff
@@ -0,0 +1,34 @@
+  href: string;
+  type: string;
+  title?: string;
+  locations?: Locations;
+  text?: LocatorText;
```
To simplify, on mobile these properties are not nullable. This way we don't have to guard for null.

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+  description?: string;
+  publisher?: Contributor;
+  imprint?: Contributor;
+  published?: string;
+  modified?: string;
```
We can use a `Date` object instead of raw strings for these.

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+  narrator?: Contributor;
+  penciler?: Contributor;
+  translator?: Contributor;
+  language?: Array<string>;
```
`languages` with an s would better reflect the fact that it's an array. And we can make it non nullable because an empty array already means no languages.

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+  readingProgression?: ReadingProgression;
+  duration?: number;
+  numberOfPages?: number;
+  abridged?: boolean;
```
I don't think `abridged` is part of the RWPM, but it could be part of the extensibility if it's needed.

We need an `otherMetadata: Map<String, Any>` to handle extensibility in the `Metadata`.

(`presentation` should be a helper using the extensibility, as well)

mickael-menu commented on r2-shared/Publication/Metadata.ts:
```diff
@@ -0,0 +1,128 @@
+    this.presentation = metadata.presentation ? new Presentation(metadata.presentation) : new Presentation({}); 
+  }
+
+  public effectiveReadingProgression(): ReadingProgression {
```
[Here are some Swift unit tests](https://github.com/readium/r2-shared-swift/blob/981594c6d935b8329a543b86b0a15a580ecc7508/r2-shared-swiftTests/Publication/MetadataTests.swift#L198-L229) for `effectiveReadingProgression()`, which is error-prone.

mickael-menu commented on r2-shared/Publication/Publication.ts:
```diff
@@ -0,0 +1,93 @@
+  public fetcher: any | null = null; // tmp
+  public services: any | null = null; // tmp
+  public format: Format | null = null;
+  public formatVersion: string | null = null;
```
We are not exposing `formatVersion` anymore (it will be soon deprecated on mobile), and I think we won't be adding `format` as well. I think it's something we need to discuss more on a Readium call.

mickael-menu commented on r2-shared/Publication/Publication.ts:
```diff
@@ -0,0 +1,93 @@
+  public formatVersion: string | null = null;
+
+  // Aliases
+  public context: Array<string> = this.manifest.context;
```
I think the `context` is not necessary in `Publication`.

mickael-menu commented on r2-shared/Publication/Publication.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+  };
+
+  public linkWithHref(href: string): Link | null {
```
[Here are some important unit tests](https://github.com/readium/r2-shared-kotlin/blob/b3e4e59745a6cb2719437e418b3f8d45d85702b4/r2-shared/src/test/java/org/readium/r2/shared/publication/PublicationTest.kt#L167-L283) to make sure that `linkwithHref()` and `linkWithRel()` variants are working properly

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+import Store from "../Store/Store";
+import { Link, Links } from "./Link";
+
+export default class PublicationManifest {
```
On mobile we recently renamed this into just `Manifest` to be more coherent with the other types (e.g. `Metadata`)

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+  public readonly resources: Links;
+  public readonly tableOfContents: Links;
+
+  public readonly manifestUrl: URL;
```
`manifestUrl` only works for WebPub right? Unless this always points to the publication URL no matter its type. What is it used for?

JayPanoz replied:
Ah yeah so every review comment about Manifest is kinda tied to discussing what happens there. 

I started from R2D2BC as I quickly noticed there was a massive overlap with NYPL original code + ours so it definitely needs feedback from other parties since it’s been probably used in a lot of places.

mickael-menu replied:
Okay we can discuss about that on the next call.

Note that it's important that `manifest` is private in `Publication`, so all these functions (if we keep them) should probably be moved to `Publication` instead of `Manifest` (which doesn't exist in R2D2BC)

aferditamuriqi replied:
Probably shouldn't be here in the first place, R2D2BC doesn't include the entire shared model as it has been defined way after R2D2BC has been developed, R2D2BC only includes the Publication and Locator model, as created before a specification was done. This models here are not to be compared or even mixed up with the R2D2BC models, but rather be created as a "new" shared models that can be added or replaces in previous implementations that haven't refactored yet to the newest definitions.

JayPanoz replied:
Thanks for the confirmation, this has been removed.

mickael-menu replied:
Great, that will simplify things 👍

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    return fetchManifest();
+  }
```
`getManifest()` and `Store` look like application logic and should probably not be part of `Shared`

JayPanoz replied:
Agreed. We’ve changed these twice within a year but see above.

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+
+  public getCoverLink(): Link | null {
+    return this.linkWithRel("cover");
+  }
```
This will soon be deprecated on mobile because we now have the `CoverService` to get the cover directly

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+      }
+    }
+    return null;
+  }
```
All of these should not be part of the shared models API, it is mostly navigator logic or redundant with other APIs such as `linkWithHref()`. Also the term "spine" is deprecated.

If some implementations rely on these APIs at the moment, I guess they could add them to the `prototype` of `Publication` in their own repos.

JayPanoz replied:
Cf. needs feedback above. 

But I can at least point out I’ve been testing extending and it should be fairly trivial to do so, without unexpected issues.

mickael-menu replied:
That's great news, thanks for testing it out

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+        }
+      }
+      return null;
+    }
```
This should be removed as it's redundant with the more generic `tableOfContents.findWithHref()`. Also I'm not quite sure why there's a fallback on the `readingOrder`, it might be implementation-specific logic and should not be part of the toolkit.

JayPanoz replied:
Yeah all the Links/Link methods aren’t used yet, as I was waiting for feedback.

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+    return null;
+  }
+
+  public getAbsoluteHref(href: string): string | null {
+    return new URL(href, this.manifestUrl.href).href;
+  }
```
The preferred way to get the absolute HREF is through `link.toURL(publication.baseURL())`

mickael-menu commented on r2-shared/Publication/PublicationManifest.ts:
```diff
@@ -0,0 +1,196 @@
+  public getRelativeHref(href: string): string | null {
+    const manifest = this.manifestUrl.href.replace("/manifest.json", ""); //new URL(this.manifestUrl.href, this.manifestUrl.href).href;
+    return href.replace(manifest, "");
+  }
```
What's the use case for this function? I feel like it should be in some kind of server handler and not here.

mickael-menu commented on r2-shared/Publication/Subject.ts:
```diff
@@ -0,0 +1,8 @@
+  name: string | LocalizedString;
+  sortAs?: string;
+  code?: string;
+  scheme?: string;
```
It's missing `links`: "Used to retrieve similar publications for the given subjects."

mickael-menu commented on r2-shared/Publication/divina/PublicationManifest.ts:
```diff
@@ -0,0 +1,14 @@
+    )
+    this.guided = manifestJSON.guided ? new Links(manifestJSON.guided) : new Links([]);
+  }
+}
```
This should not be a subclass but be implemented with the `subcollections` extensibility in `Manifest`.

See here: https://github.com/readium/r2-shared-swift/blob/981594c6d935b8329a543b86b0a15a580ecc7508/r2-shared-swift/Publication/Manifest.swift#L33

And the `PublicationCollection` class is here: https://github.com/readium/r2-shared-swift/blob/feature/publication-encapsulation/r2-shared-swift/Publication/PublicationCollection.swift

(BTW the name is not great, maybe this should be called `CoreCollection` or `JSONCollection`. This is not specific to `Publication` and could be used in other contexts, such as the OPDS registry)

JayPanoz replied:
That was on my TODO List, sorry for not having make it obvious.

mickael-menu replied:
No problem

mickael-menu commented on r2-shared/Publication/epub/Properties.ts:
```diff
@@ -0,0 +1 @@
+export type Properties = "mathml" | "onix" | "remote-resources" | "js" | "svg" | "xmp";
```
Is there a finite amount of `contains` properties? On mobile we used a simple string array, but an enum might be better if we're sure it's all there is.

JayPanoz replied:
[EPUB Spec](https://www.w3.org/publishing/epub3/epub-packages.html#app-item-properties-vocab)

[WebPub Manifest](https://github.com/readium/webpub-manifest/blob/master/extensions/epub.md#contains)

On a related note, shouldn’t the `js` value be `scripted`?

jccr replied:
In TS string literal unions are an [alternative to enums](https://2ality.com/2020/02/enum-alternatives-typescript.html#unions-of-singleton-values).

I prefer them as they are more lightweight in terms of writing them and with the final output (no runtime artifacts)

mickael-menu replied:
Looks like this is an exhaustive type indeed, let's confirm that on the call with Hadrien and probably we can migrate to enums on mobile.

JayPanoz replied:
@jccr Would you consequently prefer ReadingProgression and EPUBLayout as string literal unions too? Let’s try to keep it consistent.

mickael-menu commented on r2-shared/Publication/epub/PublicationManifest.ts:
```diff
@@ -0,0 +1,35 @@
+import { Link, Links } from "../Link";
+import { EPUBLayout } from "./Layout";
+
+export default class EPUBPublicationManifest extends PublicationManifest {
```
Same comment as for `DiViNaPublicationManifest`

mickael-menu commented on r2-shared/Publication/epub/PublicationManifest.ts:
```diff
@@ -0,0 +1,35 @@
+    } else {
+      return EPUBLayout.reflowable;
+    }
+  }
```
This should be moved to `Presentation`

JayPanoz replied:
It turns out it already is so this can be confidently removed w/o any loss.

mickael-menu replied:
Ha yes I missed that, thanks.

JayPanoz committed:
Rename PublicationManifest → Manifest

JayPanoz committed:
Turn ILink interface into LinkLike type

JayPanoz committed:
Make Links methods more idiomatic

JayPanoz committed:
Correct case of DOMRangeInfo interface

JayPanoz committed:
Change language to languages in Metadata

JayPanoz committed:
Remove format and formatVersion from Publication

JayPanoz committed:
Remove context alias in Publication

JayPanoz committed:
Rename Encrypted → Encryption

JayPanoz committed:
TODO: Update matches in MediaType and check it’s correct

JayPanoz committed:
Add MediaType prop in Link

JayPanoz committed:
Add links member in Contributor

JayPanoz committed:
Add links member in Subject

JayPanoz committed:
Add note in Divina + ePub manifests

JayPanoz committed:
Rename urlRelativeTo → toAbsoluteHREF in Link

JayPanoz committed:
Replace Format interface with type

JayPanoz committed:
Simplify toAbsoluteHREF in Link

JayPanoz committed:
Correct languages + rels array in Link

JayPanoz committed:
Correct belongsTo in Metadata

JayPanoz committed:
Add missing media types

mickael-menu commented on r2-shared/Publication/Publication.ts:
```diff
@@ -0,0 +1,87 @@
+import { Link, Links } from "./Link";
+
+export default class Publication {
+  public manifest: Manifest;
+  public fetcher: any | null = null; // tmp
+  public services: any | null = null; // tmp
```
`manifest`, `fetcher` and `services` should be private in `Publication`.

mickael-menu commented:
Here are some points worth mentioning during our next call, but feel free to discuss beforehand:

* [Should the types be normalized?](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442304815) For example, having a full–fledged `Contributor` object everywhere instead of a union type such as `type Contributor = string | IContributor | Array<string | IContributor>`.
* [Usage of interfaces to construct objects (e.g. `LinkLike`)](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442317382)
* Regarding `Locator`:
  * [Refactoring the HTML extensions (e.g. `cssSelector`) in an extension](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442337983)
  * [Making `locations` and `text` non-nullable](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442338521)
* [Using `Date` objects instead of `string` for `published` and `modified`](https://github.com/Jellybooks/jbks-shared-web/pull/1/files#r442339228)
* Broader discussion about the `Manifest`:
  * [What to do with the non-standard APIs used in R2D2BC?](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442350822) They can probably be removed from shared and [added as prototype extensions in each implementation that needs them](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442355647) 
  * [Should we remove `getManifest()` and `Store` from shared?](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442351318)
* [Should an enum be used for `contains`?](https://github.com/Jellybooks/jbks-shared-web/pull/1#discussion_r442361696)

JayPanoz committed:
Add packaged formats

JayPanoz committed:
Remove R2D2BC API from API, Store and manifestUrl

JayPanoz committed:
Make Publication members private

JayPanoz committed:
Simplify contributor

JayPanoz committed:
Simplify Subject

JayPanoz committed:
Update Locators

Make fragments, locations and text non-nullable, rename DOMRangeInfo →
DOMRange and StartEndInfo → DOMRangePoint

JayPanoz committed:
Add comments documenting interfaces, classes, their members and methods

JayPanoz committed:
Add leadingPage() to ReadingProgression

JayPanoz committed:
Correct belongsTo collection/series

JayPanoz committed:
Init JSONDictionary for subcollections

JayPanoz committed:
Init CoreCollection for extending RWPM

JayPanoz committed:
Modify handling of tableOfContents in Manifest

JayPanoz committed:
Add otherLocations in Locations Interface

JayPanoz committed:
Improve JSONDictionary

JayPanoz committed:
Switch back to original handling of tableofContents in Manifest

JayPanoz committed:
Add subcollections alias in Publication

JayPanoz committed:
Add URITemplate methods in Link

JayPanoz committed:
Add EPUB + DiViNa WebPub extensions

JayPanoz committed:
Add minimal code for Fetcher in Publication

JayPanoz committed:
Add Navigator interfaces (Public API)

JayPanoz committed:
Remove FormatLike type in Format

JayPanoz committed:
Don’t export some classes as default so that they can be augmented by modules

JayPanoz committed:
Init extensibility for link properties

JayPanoz committed:
Resolve some issues

JayPanoz committed:
Correct extension properties

JayPanoz committed:
Correct otherMetadata in Metadata

JayPanoz committed:
Switch to functions for module extensions

JayPanoz committed:
Make interfaces, exports, etc. more consistent

JayPanoz committed:
Correct ReadingProgression in VisualNavigator interface

JayPanoz committed:
Correct nullability of members in some classes

JayPanoz committed:
Add copyright header

JayPanoz committed:
Remove Fetcher and services for the time being

That should prevent confusion and services need further discussions in
the Web context.

JayPanoz committed:
Some corrections for TSCheck

JayPanoz committed:
Correct ILink properties

JayPanoz committed:
Resolve TS-error for aliases
